batch.go

```diff
@@ -0,0 +1,277 @@
+}
+
+type RequestCounts struct {
+	Processing int `json:"processing"`
+	Succeeded  int `json:"succeeded"`
+	Errored    int `json:"errored"`
+	Canceled   int `json:"canceled"`
+	Expired    int `json:"expired"`
+}
+
+func (c *Client) CreateBatch(
+	ctx context.Context,
+	request BatchRequest,
+) (*BatchResponse, error) {
+	var setters []requestSetter
+	if len(c.config.BetaVersion) > 0 {
+		setters = append(setters, withBetaVersion(c.config.BetaVersion...))
+	}
+
+	urlSuffix := "/messages/batches"
+	req, err := c.newRequest(ctx, http.MethodPost, urlSuffix, request, setters...)
+	if err != nil {
+		return nil, err
+	}
+
+	var response BatchResponse
+	err = c.sendRequest(req, &response)
+
+	return &response, err
+}
+
+func (c *Client) RetrieveBatch(
+	ctx context.Context,
+	batchId BatchId,
+) (*BatchResponse, error) {
+	var setters []requestSetter
+	if len(c.config.BetaVersion) > 0 {
+		setters = append(setters, withBetaVersion(c.config.BetaVersion...))
+	}
+
+	urlSuffix := "/messages/batches/" + string(batchId)
+	req, err := c.newRequest(ctx, http.MethodGet, urlSuffix, nil, setters...)
+	if err != nil {
+		return nil, err
+	}
+
+	var response BatchResponse
+	err = c.sendRequest(req, &response)
+
+	return &response, err
+}
+
+type BatchResult struct {
+	CustomId string `json:"custom_id"`
+	Result   struct {
+		Type    ResultType       `json:"type"`
+		Message MessagesResponse `json:"message"`
+	} `json:"result"`
+}
+
+type RetrieveBatchResponse struct {
+	httpHeader
+
+	// Each line in the file is a JSON object containing the result of a
+	// single request in the Message Batch. Results are not guaranteed to
+	// be in the same order as requests. Use the custom_id field to match
+	// results to requests.
+
+	Responses   []BatchResult
+	RawResponse []byte
+}
+
+// Untested - I don't know about this!
+func (c *Client) RetrieveBatchResults(
+	ctx context.Context,
+	batchId BatchId,
+) (*RetrieveBatchResponse, error) {
+	var setters []requestSetter
+	if len(c.config.BetaVersion) > 0 {
+		setters = append(setters, withBetaVersion(c.config.BetaVersion...))
+	}
+
+	urlSuffix := "/messages/batches/" + string(batchId) + "/results"
+	req, err := c.newRequest(ctx, http.MethodGet, urlSuffix, nil, setters...)
+	if err != nil {
+		return nil, err
+	}
+
+	var response RetrieveBatchResponse
+
+	res, err := c.config.HTTPClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	response.SetHeader(res.Header)
+
+	if err := c.handlerRequestError(res); err != nil {
+		return nil, err
+	}
+
+	response.RawResponse, err = io.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// for each line in the response, decode the JSON object into a MessagesResponse
+	// and append it to the Responses slice.
+	// this is tricky because the content within the response may contain newline control characters (\n)
+	// TODO: test this and make sure it works
+	for _, line := range bytes.Split(response.RawResponse, []byte("\n")) {
```

Comment:
todo: look at this again. not overly convinced with this logic...

Reply:
I'm now convinced by this logic. Evidence: https://goplay.tools/snippet/tDPm3GJVv0_s